File: frontend/src/app/core/widget/widget-snapshot.ts
import { TodayNutritionSummary } from '../data/today-nutrition.service';

/**
 * documentation/Features/Android kezdőképernyő widget.md — the JSON contract between the Angular
 * `WidgetSnapshotService` (writer) and the native `AppWidgetProvider`s (readers). Written to the
 * `@capacitor/preferences` / `CapacitorStorage` key below, exactly like `lm2_notifBgPlan`. Every
 * user-facing string is pre-localized here (the widget process runs no JS); the native side only
 * formats numbers and assembles `"<intake> / <goal>"` lines.
 */
export const WIDGET_SNAPSHOT_KEY = 'lm2_widgetSnapshot';
export const WIDGET_SNAPSHOT_VERSION = 1;

/** Route deep-links a widget tap stashes for `NotificationSchedulerService.drainPendingRoute()`. */
export const WIDGET_ROUTES = {
  open: '/tabs/home',
  nutrition: '/tabs/food',
  steps: '/tabs/menu/steps',
  profile: '/tabs/menu/profile',
  newMeal: '/tabs/food/meal/new',
  newClimb: '/tabs/workout/climbing',
} as const;

/** i18n keys the service resolves via `TranslateService.instant` before calling {@link buildWidgetSnapshot}. */
export const WIDGET_LABEL_KEYS = {
  appTitle: 'WIDGET.APP_TITLE',
  nutritionTitle: 'WIDGET.NUTRITION_TITLE',
  stepsTitle: 'WIDGET.STEPS_TITLE',
  quickTitle: 'WIDGET.QUICK_TITLE',
  kcal: 'WIDGET.KCAL',
  protein: 'WIDGET.PROTEIN',
  carbs: 'WIDGET.CARBS',
  fat: 'WIDGET.FAT',
  stepsUnit: 'WIDGET.STEPS_UNIT',
  remaining: 'WIDGET.REMAINING',
  loggedOut: 'WIDGET.LOGGED_OUT',
  noProfile: 'WIDGET.NO_PROFILE',
  newMeal: 'WIDGET.NEW_MEAL',
  newClimb: 'WIDGET.NEW_CLIMB',
} as const;

export type WidgetLabels = Record<keyof typeof WIDGET_LABEL_KEYS, string>;

export interface WidgetNutrient {
  intake: number;
  goal: number;
}

export interface WidgetSnapshot {
  version: number;
  writtenAt: number;
  loggedIn: boolean;
  lang: string;
  /** `null` when the profile lacks the TDEE inputs — the nutrition / summary widgets show `labels.noProfile`. */
  nutrition: {
    incomplete: boolean;
    kcal: WidgetNutrient;
    proteinG: WidgetNutrient;
    carbsG: WidgetNutrient;
    fatG: WidgetNutrient;
  } | null;
  steps: {
    count: number;
    goal: number;
  };
  labels: WidgetLabels;
  routes: typeof WIDGET_ROUTES;
}

export interface WidgetSnapshotInputs {
  loggedIn: boolean;
  lang: string;
  nutrition: TodayNutritionSummary;
  stepCount: number;
  stepGoal: number;
  labels: WidgetLabels;
  now?: number;
}

const clampNonNegative = (value: number): number => (Number.isFinite(value) && value > 0 ? Math.round(value) : 0);

export function buildWidgetSnapshot(inputs: WidgetSnapshotInputs): WidgetSnapshot {
  const n = inputs.nutrition;
  return {
    version: WIDGET_SNAPSHOT_VERSION,
    writtenAt: inputs.now ?? Date.now(),
    loggedIn: inputs.loggedIn,
    lang: inputs.lang,
    nutrition:
      inputs.loggedIn && n.computable
        ? {
            incomplete: n.incomplete,
            kcal: nutrient(n.kcal),
            proteinG: nutrient(n.proteinG),
            carbsG: nutrient(n.carbsG),
            fatG: nutrient(n.fatG),
          }
        : null,
    steps: {
      count: clampNonNegative(inputs.stepCount),
      goal: clampNonNegative(inputs.stepGoal),
    },
    labels: inputs.labels,
    routes: WIDGET_ROUTES,
  };
}

function nutrient(progress: { intake: number; goal: number }): WidgetNutrient {
  return { intake: clampNonNegative(progress.intake), goal: clampNonNegative(progress.goal) };
}

File: frontend/android/app/src/main/java/hu/bumler/lm2/widget/WidgetSnapshot.kt
package hu.bumler.lm2.widget

import android.content.Context
import org.json.JSONObject

/**
 * documentation/Features/Android kezdőképernyő widget.md — the parsed form of the `lm2_widgetSnapshot`
 * JSON that `WidgetSnapshotService` writes into the `CapacitorStorage` SharedPreferences file
 * (`@capacitor/preferences`). Same bridge the notification background worker uses for `lm2_notifBgPlan`.
 * All user-facing strings are already localized on the JS side; the widget only formats numbers.
 * TS contract: frontend/src/app/core/widget/widget-snapshot.ts.
 */
data class WidgetSnapshot(
    val loggedIn: Boolean,
    /** `null` => the profile lacks TDEE inputs; the nutrition / summary widgets show `noProfile`. */
    val nutrition: Nutrition?,
    val stepCount: Int,
    val stepGoal: Int,
    private val labels: Map<String, String>,
    private val routes: Map<String, String>,
) {
    data class Nutrient(val intake: Int, val goal: Int)

    data class Nutrition(
        val incomplete: Boolean,
        val kcal: Nutrient,
        val protein: Nutrient,
        val carbs: Nutrient,
        val fat: Nutrient,
    )

    fun label(key: String): String = labels[key].orEmpty()

    fun route(key: String): String = routes[key] ?: DEFAULT_ROUTE

    companion object {
        const val PREFS_NAME = "CapacitorStorage"
        const val SNAPSHOT_KEY = "lm2_widgetSnapshot"
        private const val DEFAULT_ROUTE = "/tabs/home"

        // Label keys — must match WIDGET_LABEL_KEYS in widget-snapshot.ts.
        const val L_APP_TITLE = "appTitle"
        const val L_NUTRITION_TITLE = "nutritionTitle"
        const val L_STEPS_TITLE = "stepsTitle"
        const val L_QUICK_TITLE = "quickTitle"
        const val L_KCAL = "kcal"
        const val L_PROTEIN = "protein"
        const val L_CARBS = "carbs"
        const val L_FAT = "fat"
        const val L_STEPS_UNIT = "stepsUnit"
        const val L_REMAINING = "remaining"
        const val L_LOGGED_OUT = "loggedOut"
        const val L_NO_PROFILE = "noProfile"
        const val L_NEW_MEAL = "newMeal"
        const val L_NEW_CLIMB = "newClimb"

        // Route keys — must match WIDGET_ROUTES in widget-snapshot.ts.
        const val R_OPEN = "open"
        const val R_NUTRITION = "nutrition"
        const val R_STEPS = "steps"
        const val R_PROFILE = "profile"
        const val R_NEW_MEAL = "newMeal"
        const val R_NEW_CLIMB = "newClimb"

        fun read(context: Context): WidgetSnapshot? {
            val raw = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
                .getString(SNAPSHOT_KEY, null) ?: return null
            return try {
                parse(JSONObject(raw))
            } catch (e: Exception) {
                null
            }
        }

        private fun parse(root: JSONObject): WidgetSnapshot {
            val nutritionJson = root.optJSONObject("nutrition")
            val steps = root.optJSONObject("steps")
            return WidgetSnapshot(
                loggedIn = root.optBoolean("loggedIn", false),
                nutrition = nutritionJson?.let {
                    Nutrition(
                        incomplete = it.optBoolean("incomplete", false),
                        kcal = nutrient(it.optJSONObject("kcal")),
                        protein = nutrient(it.optJSONObject("proteinG")),
                        carbs = nutrient(it.optJSONObject("carbsG")),
                        fat = nutrient(it.optJSONObject("fatG")),
                    )
                },
                stepCount = steps?.optInt("count", 0) ?: 0,
                stepGoal = steps?.optInt("goal", 0) ?: 0,
                labels = stringMap(root.optJSONObject("labels")),
                routes = stringMap(root.optJSONObject("routes")),
            )
        }

        private fun nutrient(obj: JSONObject?): Nutrient =
            Nutrient(obj?.optInt("intake", 0) ?: 0, obj?.optInt("goal", 0) ?: 0)

        private fun stringMap(obj: JSONObject?): Map<String, String> {
            if (obj == null) return emptyMap()
            val out = HashMap<String, String>()
            for (key in obj.keys()) {
                out[key] = obj.optString(key, "")
            }
            return out
        }
    }
}
